### DISCUSSION TITLE
Ignoring the Typeshed stubs for the library I'm working on

### DISCUSSION THREAD

**Discussion by richardm-stripe** [microsoft/pyright#5585] [Q&A] [answered]
I'm currently experimenting with adding type annotations to a library ([stripe-python](https://github.com/stripe/stripe-python)) that has [stubs defined in Typeshed](https://github.com/python/typeshed/tree/main/stubs/stripe).

While running `pyright` locally, I would like to completely ignore the definitions in Typeshed, and do type analysis based only on the local inline annotations for the package in my workspace.

Currently, I can only accomplish this by forking Typeshed, `rm -rf`ing the stubs, and using `--typeshed-path` to point at my fork. I'm not entirely satisfied with this because I'm working on a team and pointing at a local fork is not really a configuration setting that I can "commit to the repository", so to speak (unless I vendor the entirety of Typeshed).

I hoped that adding a `py.typed` file to the root of my project would cause the inline `*.py` annotations to take precedence over Typeshed, as it does when resolving libraries underneath `lib/site-packages` but it appears to ignored in this case.

Other things I tried: 
  * put an empty directory named after my package locally in the project beneath `typings`, hoping to "shadow" Typeshed. 
  * setting typeshedPath to `$HOME/typeshed` so my collaborators would at least have a standard place to put their fork
  * looked for a setting to "skip" a particular package's Typeshed stubs

Has anybody else been in this situation and have advice? 

Something else I've noticed is that I can't do `pyright --verifytypes .` in order to get information about the type completeness of the package as it exists in my workspace. I can only do this from a project that has my package installed as a dependency. This makes me wonder: am I just approaching this entirely wrong & should I be setting my workspace/environment up completely differently?

(P.S. just want to say, thank you for this tool! So far I have been having a very good experience and am blown away by the thoughtfulness of the design and especially the documentation & the quality of the responses in this forum.)

**Comment by erictraut** [comment, marked as answer]
It's great to hear that you're looking at adding inlined type annotations to `stripe-python`. That should help improve the experience for consumers of this library.

For information about pyright's import resolution order, refer to [this documentation](https://microsoft.github.io/pyright/#/import-resolution?id=resolution-order). These rules are largely dictated by [PEP 561](https://peps.python.org/pep-0561/), and pyright is compliant with this PEP.

I think the part that you're missing is that you need to do an editable install of your library (`pip install -e .`) so it acts as though it has been installed within your python environment. Depending on the setup mechanism you use, you may run into issues with PEP 660, which can cause problems for static analysis tools. For workarounds, refer to [this documentation](https://microsoft.github.io/pyright/#/import-resolution?id=editable-installs). 

Once your library is installed, it should override any bundled typeshed stubs. You should also be able to use `--verifytypes`.

**Reply by richardm-stripe** [reply]
Hello. A huge belated thank you for this! The editable install tip was exactly what I was missing. 


I'm retrospecting a little bit now the annotations are [out](https://github.com/python/typeshed/issues/10966). When I *don't* do an editable install, I'm struggling a little bit to reconcile the description [in the docs](https://microsoft.github.io/pyright/#/import-resolution?id=resolution-order) with some of the behavior I'm seeing. Specifically, the order is supposed to be

> (2) Try to resolve using code within the workspace.
> (5) Try to resolve using a third-party typeshed stub.

But when I created a [test workspace](https://github.com/richardm-stripe/pyright-demo/tree/main/stripe) with files

```shell
# Structure
stripe/demo.py
stripe/util.py # file with the same name exists in Typeshed
stripe/not_in_typeshed.py
```

```python
# demo.py
from stripe.util import notInTypeshed
from stripe.not_in_typeshed import veryNotInTypeshed
```

I expect both `notInTypeshed` and `veryNotInTypeshed` to resolve to their definitions in the workspace. What I actually observe is that `notInTypeshed` doesn't resolve because `stripe.util` resolves to the module in Typeshed.

It's not really a problem for me as the editable install works but I would like to better understand how to think about what is going on here.

**Reply by erictraut** [reply]
Congrats on getting the annotations in place for the `stripe-python` library.

There are two issues at play with the scenario you outlined above:
1. You haven't included a `stripe/__init__.py` file, so `stripe` is not a traditional package. Instead, it's a namespace package. The Python import resolver always prefers traditional packages. Since the typeshed module is a traditional package, it is preferred.
2. There is a bug in pyright's import resolver. Even if you add the `stripe/__init__.py` file, it still prefers the typeshed stub over the local file when it shouldn't. I've created a [new bug report](https://github.com/microsoft/pyright/issues/6406) to track this issue.

**Reply by erictraut** [reply]
This bug will be fixed in the next release.
